rfcs/rework-index-updates.md

```diff
@@ -0,0 +1,60 @@
+# Summary
+
+This RFC proposes moving away from the current model of fetching new releases
+to build, moving from running [crates-index-diff] in a thread to using webhooks
+and the [crates-index] crate.
+
+# Motivation
+
+While the current approach has worked well for us so far, it has some problems:
+
+* Running the update in a cronjob every 2 minutes is wasteful, as there is
+  often a greater delay between two publishes.
+* Running the update in a crobjob every 2 minutes adds delay to getting the
+  documentation built if the queue is empty, as the release might potentially
+  have to wait those extra two minutes.
+* The approach doesn't scale, if we want to move to a setup where there is more
+  than a single frontend server we'd have to elect which server runs the fetch.
+* The way crates-index-diff stores its state (a branch in the local repo) is
+  fragile, as it might become out of sync causing the loss of a publish.
+* The way crates-index-diff stores its state makes it hard to move the server
+  installation, as the index repository needs to be moved as well.
+
+# Proposal
+
+We configure the `crates.io-index` repository to send a webhook to a new
+endpoint, `/_/index-webhook`, which starts a index sync in the background. The
+payload of the webhook is ignored, but the webhook signature is validated if a
+secret key is provided to the application through an environment variable.
+
+When an index synchronization starts, the [crates-index] crate is used to load
+in memory a list of all crates, their versions and whether each version is
+yanked. Then, the full list of releases and queued crates is fetched from the
+database, and it's compared with the contents of the index. Finally, idempotent
+queries are sent to the database to update its state (queueing crates and
+changing the yanked status) where needed.
```

Comment:
This sounds really expensive. We currently have `238677` crates in the database and just running `select name, version, yanked from crates inner join releases on crates.id = releases.crate_id` takes a full second in production. Comparing it to the index sounds even slower. I would rather have the webhook contain the metadata of the crate that's been uploaded and separate the webhook from synchronization. Synchronization is still a good idea, I just don't think it should happen on every release (maybe every week or so?).

Possible alternatives to full synchronization:
- The database is the one true source of knowledge and if we crash while processing a webhook, the crate will be lost until we do another full synchronization. I don't expect crashes to be common, so as long as we have _any_ synchronization I think this is reasonable.
- The webhook has some sort of two-way communication and resends the payload until it gets a confirmation that it was received/committed
- The above, but the confirmation is 'docs.rs closed the HTTP connection' and we make sure we write the payload to disk before we do so.

Reply:
> Comparing it to the index sounds even slower.

On my machine, collecting all the crates in the index takes only 2 seconds [(source)](https://gist.github.com/pietroalbini/6e069454498d5028dc19d5ea4491138f). Considering how rarely crates are published I think this is not so bad, and I feel like the advantages outweight the performance issues.

> The database is the one true source of knowledge and if we crash while processing a webhook, the crate will be lost until we do another full synchronization. I don't expect crashes to be common, so as long as we have any synchronization I think this is reasonable.

This is the status quo, right?

> I would rather have the webhook contain the metadata of the crate that's been uploaded and separate the webhook from synchronization.
> * The webhook has some sort of two-way communication and resends the payload until it gets a confirmation that it was received/committed
> * The above, but the confirmation is 'docs.rs closed the HTTP connection' and we make sure we write the payload to disk before we do so.

We don't control GitHub webhooks, and we can't really do this.

Reply:
> On my machine, collecting all the crates in the index takes only 2 seconds (source). Considering how rarely crates are published I think this is not so bad, and I feel like the advantages outweight the performance issues.

Hmm, in the common case that doesn't seem outrageous, but I think it will slow down the server a lot when we get 300 crates from the same project published in a row. I'm also not convinced that the advantages are that much greater than running a full synchronization on a timer occasionally - do we really need to make sure we build _every_ crate _immediately_ after publication, even if the site just went down?

> > The database is the one true source of knowledge and if we crash while processing a webhook, the crate will be lost until we do another full synchronization. I don't expect crashes to be common, so as long as we have any synchronization I think this is reasonable.

> This is the status quo, right?

With two major differences:
1. Updates are push instead of poll, so we no longer run `git fetch` on the index every minute
2. We have full synchronization on a timer, so that crates will no longer be lost if we miss an update somehow.

> We don't control GitHub webhooks, and we can't really do this.

That's a shame, but that makes sense.